toolchain/lower/file_context.cpp

```diff
@@ -116,6 +121,166 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
+                                   equivalent_entry.first);
+          });
+
+          // Removed the replaced specific from the list of emitted specifics.
+          specifics[j] = specifics[specifics.size() - 1];
+          specifics.pop_back();
+          --j;
+        } else {
+          // Only mark non-equivalence based on state for starting specifics.
+          non_equivalent_specifics.Insert(test_equivalence);
+        }
+      }
+    }
+  });
+}
+
+auto FileContext::ReplaceSpecificWithSpecific(SemIR::SpecificId to_replace,
+                                              SemIR::SpecificId replace_with)
+    -> void {
+  specific_functions_[to_replace.index]->replaceAllUsesWith(
+      specific_functions_[replace_with.index]);
+}
+
+auto FileContext::DeleteFunctionSpecific(SemIR::SpecificId to_replace,
+                                         SemIR::SpecificId replace_with)
+    -> void {
+  specific_functions_[to_replace.index]->eraseFromParent();
+  specific_functions_[to_replace.index] =
+      specific_functions_[replace_with.index];
+}
+
+auto FileContext::CheckTypeEquivalence(SemIR::SpecificId specific_id1,
+                                       SemIR::SpecificId specific_id2) -> bool {
+  CARBON_CHECK(specific_id1.has_value() && specific_id2.has_value());
+  auto specific1_entry = lowered_specifics_types.Lookup(specific_id1);
+  auto specific2_entry = lowered_specifics_types.Lookup(specific_id2);
+  CARBON_CHECK(specific1_entry && specific2_entry,
+               "Type entries expected for specifics");
+  return specific1_entry.value().Equals(specific2_entry.value());
+}
+
+// TODO: change recursion to a worklist
+// NOLINTNEXTLINE(misc-no-recursion)
+auto FileContext::CheckStateEquivalence(
+    SemIR::SpecificId specific_id1, SemIR::SpecificId specific_id2,
+    Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>&
+        visited_equivalent_specifics) -> bool {
+  auto specific1_entry = lowered_states.Lookup(specific_id1);
+  auto specific2_entry = lowered_states.Lookup(specific_id2);
+  CARBON_CHECK(specific1_entry && specific2_entry,
+               "State entries expected for specifics");
+  llvm::SmallVector<LoweringFunctionState>& states1 = specific1_entry.value();
+  llvm::SmallVector<LoweringFunctionState>& states2 = specific2_entry.value();
+
+  if (states1.size() != states2.size()) {
+    return false;
+  }
+  std::pair<SemIR::SpecificId, SemIR::SpecificId> outer_pair = {specific_id1,
+                                                                specific_id2};
+  for (unsigned i = 0; i < states1.size(); ++i) {
+    auto& state1 = states1[i];
+    auto& state2 = states2[i];
+    if (state1.inst_id_index != state2.inst_id_index ||
+        state1.global_values.size() != state2.global_values.size() ||
+        state1.calls.size() != state2.calls.size() ||
+        state1.types.size() != state2.types.size()) {
+      return false;
+    }
+    for (unsigned i = 0; i < state1.global_values.size(); ++i) {
```

Comment:
Could this use `llvm::ArrayRef::equals`?

https://llvm.org/doxygen/classllvm_1_1ArrayRef.html#a21ff1fd12b25b79fcd4449c35dc814b6